portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -127,18 +130,59 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     
     ArrayList<Comment> comments = new ArrayList<Comment>();
     int numberToDisplay = getNumberToDisplay(request); 
+    String paginationInstruction = (String) request.getParameter("pageAction");
 
     results.asIterable().forEach(comment -> {
       comments.add(Comment.fromEntity(comment));
     });
+    int[] commentRange = getRangeOfCommentsToDisplay(paginationInstruction, 
+      numberToDisplay, comments.size());
     List<Comment> commentsToDisplay = 
-      comments.subList(0, Math.min(numberToDisplay, comments.size()));
+      comments.subList(commentRange[0], commentRange[1]);
+    currentIndexOfFirstComment = commentRange[0];
 
     String json = convertToJson(commentsToDisplay); 
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
+  /** Returns the range of comments that should be displayed on
+    * the site, given the user's instructions (reflecting
+    * whether to jump forward a page, jump backwards, or
+    * stay on the same page of comments).
+    * @param instruction Of the form ""next"", ""previous"", 
+        or ""none"".
+    * @param numberToDisplay Indicates the desired number of comments
+        on the page.
+    * @param totalNumberComments Indicates the number of comments currently
+        in the database.
+    * @return An array of the form `[startIndex, stopIndex + 1]`, where
+        startIndex is the index of the first comment that should be displayed
+        and stopIndex is the index of the last comment that should be
+        displayed. */
+  private int[] getRangeOfCommentsToDisplay(String instruction, 
```

Comment:
Optional: the Range class as an optimization. 

https://github.com/google/guava/wiki/RangesExplained

Reply:
Great idea! I've added it in